### ISSUE TITLE
`dsolve` output/return value is unclear

### ISSUE THREAD

**Issue by karlicoss** [sympy/sympy#15707]
I ran into a bug while solving a system of odes, not covered by any of unit tests. While trying to fix it, I realised I don't understand what `dsolve` actually returns -- lots of existing code just assumes it returns a single solutions, however that doesn't always seem to be the case.

First, the [documentation](https://github.com/sympy/sympy/blob/master/sympy/solvers/ode.py#L533-L537) claims:
```
:py:meth:`~sympy.solvers.ode.dsolve` always returns an
          :py:class:`~sympy.core.relational.Equality` class (except for the
          case when the hint is ``all`` or ``all_Integral``).  If possible, it
          solves the solution explicitly for the function being solved for.
          Otherwise, it returns an implicit solution.    
```
However, that's not always true: if you try that `dsolve(Eq(diff(f(x), x), x / f(x)))`, you get two solutions: `sqrt(C + x^2)` and `-sqrt(C + x^2)`. Is it the specific solver's fault? If it is, what is it meant to do, return an arbitrary solution? That sounds pretty wrong.

Next, it's unclear what should be the return value for the systems of ODE. In most cases currently it is a list, where solutions are orderered according to the `func` argument. However, some solvers [return a set ](https://github.com/sympy/sympy/blob/54d7726c182065fc25df69b5d1030e568931d82f/sympy/solvers/ode.py#L8336) (I believe that's just a bug in implementation), and again, it's unclear what happens if the system admits multiple solutions.

Another thing is: if by default `dsolve` returns a single ('default') solution to ODE, and we use `dsolve` within solvers internally, wouldn't it potentially lose some of the solutions? Can't come up to the example/test quickly, but maybe I'm just missing something. Perhaps `dsolve` is only meant for users of the module and developers should instead use some other internal function with standartized output that always returns a list of solutions? 

Also somewhat related: https://github.com/sympy/sympy/issues/8006

**Comment by oscarbenjamin** [comment]
It may not be well documented but the current intention is that `dsolve` for a single ODE should either return an `Eq`, or a list of `Eq`s if multiple solutions are found. Ideally the `Eq`s are in the standard form `f(x)=...` but it's not always possible to put them in that form.

To use `dsolve` consitently you'd need to do something like:
```python
sol = dsolve(...)
if isinstance(sol, Eq):
    # handle single solution
elif isinstance(sol, list):
    # handle multiple
else:
    # shouldn't happen...
```
(Probably a better API would just return a list every time)

In some cases for a single ODE `dsolve` chooses a method that finds some but not all solutions. It's certainly not guaranteed right now that all solutions will be returned although the intention is to improve it as far as possible in that direction.

For systems of ODEs the code is not very complete and likely inconsistent.

When does `dsolve` return a set?

**Comment by asmeurer** [comment]
The other possibility is NotImplementedError. We could add some flag to make it always return a list, similar to the flags that exist for `solve`. 

If a solver is returning a `set` that's a bug.

**Comment by ShubhamKJha** [comment]
I would like to work on it.

**Comment by oscarbenjamin** [comment]
Great! The thing it's not entirely clear what is to be done here.

I also find it annoying that `dsolve` sometimes returns a list and sometimes a single solution. However it's been that way for some time and 99% of the time it returns a single solution so that's what most users are expecting. We can't change that - it's too big a break in backwards compatibility.

@asmeurer's suggestion to use a flag sounds useful but I haven't looked into it myself.

Another possibility is to improve the docs so that this is less surprising to users.

It might be worth having a trawl of the SymPy code to see how different parts of the codebase call dsolve and how they handle this. `grep` shows me which files contain "dsolve" but it might not be very useful:
```console
$ grep -r --include=\*.py -l dsolve sympy
sympy/solvers/tests/test_constantsimp.py
sympy/solvers/tests/test_ode.py
sympy/solvers/tests/test_pde.py
sympy/solvers/__init__.py
sympy/solvers/solvers.py
sympy/solvers/pde.py
sympy/solvers/ode.py
sympy/solvers/deutils.py
sympy/utilities/tests/test_wester.py
sympy/physics/continuum_mechanics/beam.py
```

**Comment by ShubhamKJha** [comment]
Change of docstrings seems to be more of a useful hack, given that users might have used dsolve for both a single solution and a list of solutions. I will look into the implementation of ```dsolve``` and check for better option.

**Comment by ShubhamKJha** [comment]
@asmeurer It appears that some of the `solvefunc` functions are returning `set`. 

```
In [74]: eq
Out[74]:
                     d            2             d
[-x(t)*y(t)*sin(t) + --(x(t)), - y (t)*sin(t) + --(y(t))]
                     dt                         dt

In [75]: dsolve(eq)
Out[75]:
               C1
             -e                     -1
{x(t) = ---------------, y(t) = -----------}
            C1                  C1 - cos(t)
        C2*e   - cos(t)
```
If `set `is unwanted I would try to return `list `values or do you have any suggestions on what should be the standard return type.

**Comment by oscarbenjamin** [comment]
So that's a system of ODEs. As I said above some of those cases might be inconsistent but that should be fixed. That case is different because unlike the single ODE case we have to provide a container of some sort because the solution will never be a single eq. It could be a list or a set but we should make it consistent.

In future can you provide code that can be easily copied and tests? E.g.:
```python
from sympy import *
x, y = symbols('x y', cls=Function)
t = Symbol('t')
eq = [x(t).diff(t) - x(t)*y(t)*sin(t), y(t).diff(t) - y(t)**2*sin(t)]
print(dsolve(eq))

from sympy.solvers.ode import classify_sysode
pprint(classify_sysode(eq))
```
Showing an isympy session is fine if you also show the code needed to reproduce the output.

For this case classify_sysode shows:
```
{'no_of_equation': 2, 'eq': [-x(t)*y(t)*sin(t) + Derivative(x(t), t), -y(t)**2*sin(t) + Derivative(y(t), t)], 'func': [x(t), y(t)], 'order': {y(t): 1, x(t): 1}, 'func_coeff': {(0, x(t), 0): -y(t)*sin(t), (0, x(t), 1): 1, (0, y(t), 0): -x(t)*sin(t), (0, y(t), 1): 0, (1, x(t), 0): 0, (1, x(t), 1): 0, (1, y(t), 0): 0, (1, y(t), 1): 1}, 'is_linear': False, 'type_of_equation': 'type4'}
```

So the relevant solver will be 2eq order1 nonlinear and type 4. You can see it here:
[https://github.com/sympy/sympy/blob/673eac7fdf3a57be7ea907a822e063baae0c3158/sympy/solvers/ode.py#L8330](https://github.com/sympy/sympy/blob/673eac7fdf3a57be7ea907a822e063baae0c3158/sympy/solvers/ode.py#L8330)

The solver below also returns a set but the one below that a list

If you want to address this for systems of ODEs you should first do a survey. There are many of these methods. Which ones return lists and which ones return sets?

See also #15995

**Comment by ShubhamKJha** [comment]
> In future can you provide code that can be easily copied and tests?

I will.

> If you want to address this for systems of ODEs you should first do a survey

I am on it. Just there are so many. 
One thing that came to my mind was to have a Data Structure(class) that can give individual results as well as lists (it has `getitem `method) So a standard output would be resulted.

**Comment by oscarbenjamin** [comment]
I don't think we want to change the return type except as needed to make it consistent. If most cases currently return lists and a couple return sets then we should change it so that they all return lists. 

Consistency is more important than the actual return type here. If a user expects to be able to do:
```python
sol = dsolve(...)
print(sol[0])
```
then their code will break if dsolve unexpectedly returns a set.

You should also check the docstring for dsolve and the docs for the ODE module to see exactly what is said about the types returned by dsolve.
